### PR TITLE
DALA-1773 Patch Company Mechanism

### PR DESCRIPTION
**Pull request by MarcTM01**
# Pull Request \<Title>
`<Description here>`
## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described in the JIRA-Issue
- [x] At least one E2E Test exists testing the new feature
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] The new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on 
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] The new feature is manually used/tested/observed on dev server
  - [x] All implemented Social Logins have been tested manually in the UI
- [x] If any work on the UI is to be merged, those changes were also documented in the Figma

### CODE REVIEW

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/CompanyApi.kt [hunk 1]
```diff
@@ -181,9 +184,63 @@ interface CompanyApi {
...
 
+    /**
+     * A method to update company informtion for one specific company identified by its company Id
+     * @param companyId identifier of the company in dataland
+     * @param companyInformation includes the company information
+     * @return updated information about the company
+     */
+    @Operation(
+        summary = "Update company information selectively",
+        description = "Changed elements of a company information behind the given company Id is updated.",
```
Comment (ahoecherl) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/CompanyApi.kt:198]:
```suggestion
        description = "Provided fields of the company associated with the given company Id are updated.",
```

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/CompanyApi.kt [hunk 2]
```diff
@@ -181,9 +184,63 @@ interface CompanyApi {
...
+
+    /**
+     * A method to update company informtion entirely
+     * @param companyId identifier of the company in dataland
+     * @param companyInformation includes the company information
+     * @return updated information about the company
+     */
+    @Operation(
+        summary = "Update company information entirely",
+        description = "all elements of a company information behind the given company Id is updated.",
```
Comment (ahoecherl) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/CompanyApi.kt:225]: Replace all company information of the company associated with the given company Id

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/repositories/CompanyIdentifierRepository.kt [hunk 3]
```diff
@@ -2,9 +2,27 @@ package org.dataland.datalandbackend.repositories
...
 import org.dataland.datalandbackend.entities.CompanyIdentifierEntityId
+import org.dataland.datalandbackend.entities.StoredCompanyEntity
+import org.dataland.datalandbackend.model.enums.company.IdentifierType
 import org.springframework.data.jpa.repository.JpaRepository
 
 /**
  * A JPA interface for accessing the CompanyIdentifier Entity
  */
-interface CompanyIdentifierRepository : JpaRepository<CompanyIdentifierEntity, CompanyIdentifierEntityId>
+interface CompanyIdentifierRepository : JpaRepository<CompanyIdentifierEntity, CompanyIdentifierEntityId> {
```
Comment (ahoecherl) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/repositories/CompanyIdentifierRepository.kt:12]: What is the reason that these two functions have appeared now as part of the PUT/PATCH story?

Comment (emanueldfine) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/repositories/CompanyIdentifierRepository.kt:12]: The `deleteAllByCompanyAndIdentifierType()` is used to delete all identifiers of the company for a specific identifier type (e.g. all Leis) from the database before writing the "updated" Leis to the database.
It is used for the `PATCH` endpoint. You can see it kind of like a PUT on a identifier-type-level.

The `deleteAllByCompany()` is used to delete all identifiers of the company from the database before writing the "updated" identifiers to the database.
It is used for the `PUT` endpoint to erase all existing identifiers, before writing the new ones.

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/CompanyApi.kt (hunk 1 above)
Comment (emanueldfine) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/CompanyApi.kt]: Agreed, changend and pushed

**Pull request merged**